Objective-C/Tests/ReplicatorTest.m

```diff
@@ -1855,6 +1855,131 @@ - (void) testPushAndForget {
+    Assert([otherDB saveDocument: doc1 error: &error]);
+    
+    CBLMutableDocument* doc2 = [[CBLMutableDocument alloc] initWithID: @"pass"];
+    [doc2 setString: @"pass" forKey: @"name"];
+    Assert([otherDB saveDocument: doc2 error: &error]);
+    
+    // replicator with pull filter
+    NSMutableSet<NSString*>* docIds = [NSMutableSet set];
+    id target = [[CBLDatabaseEndpoint alloc] initWithDatabase: otherDB];
+    CBLReplicatorConfiguration* config = [self configWithTarget: target
+                                                           type: kCBLReplicatorTypePull
+                                                     continuous: isContinuous];
+    config.pullFilter = ^BOOL(CBLDocument* document, CBLDocumentFlags flags) {
+        AssertNotNil(document.id);
+        [docIds addObject: document.id];
+        
+        BOOL isDeleted = (flags & kCBLDocumentFlagsDeleted) == kCBLDocumentFlagsDeleted;
+        if (isDeleted) {
+            // if deleted only allow  `docID = pass` is allowed.
+            return [document.id isEqualToString: @"pass"];
+        }
+        // allow all docs with `name = pass`
+        return [[document stringForKey: @"name"] isEqualToString: @"pass"];
+    };
+    
+    [self run: config errorCode: 0 errorDomain: nil];
+    
+    // Update the `_removed` flag
+    doc1 = [[otherDB documentWithID: @"doc1"] toMutable];
+    [doc1 setData: @{@"_removed": @YES}];
+    Assert([otherDB saveDocument: doc1 error: &error]);
+    
+    doc2 = [[otherDB documentWithID: @"pass"] toMutable];
+    [doc2 setData: @{@"_removed": @YES}];
+    Assert([otherDB saveDocument: doc2 error: &error]);
+    
+    // pull replication again...
+    [self run: config errorCode: 0 errorDomain: nil];
+    
+    AssertEqual(docIds.count, 2u);
+    
+    // shouldn't delete the one with `docID != pass`
+    AssertNotNil([self.db documentWithID: @"doc1"]);
+    AssertNil([self.db documentWithID: @"pass"]);
+}
+
+
+- (void) testPushRemovedDocWithFilter: (BOOL)isContinuous {
+    NSError* error;
+    CBLMutableDocument* doc1 = [[CBLMutableDocument alloc] initWithID: @"doc1"];
+    [doc1 setString: @"pass" forKey: @"name"];
+    Assert([self.db saveDocument: doc1 error: &error]);
+    
+    CBLMutableDocument* doc2 = [[CBLMutableDocument alloc] initWithID: @"pass"];
+    [doc2 setString: @"pass" forKey: @"name"];
+    Assert([self.db saveDocument: doc2 error: &error]);
+    
+    // replicator with pull filter
+    NSMutableSet<NSString*>* docIds = [NSMutableSet set];
+    id target = [[CBLDatabaseEndpoint alloc] initWithDatabase: otherDB];
+    CBLReplicatorConfiguration* config = [self configWithTarget: target
+                                                           type: kCBLReplicatorTypePush
+                                                     continuous: isContinuous];
+    config.pushFilter = ^BOOL(CBLDocument* document, CBLDocumentFlags flags) {
+        AssertNotNil(document.id);
+        [docIds addObject: document.id];
```

Comment:
I think the correct logic should be adding the docIds only when getting the access removed flag?